tr2: add 60 fps

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This PR adds support for 60 FPS. Since TR2 currently doesn't have a dedicated item in the inventory ring, the option to control this was added to the config tool intead.

This is just a draft that I wanted to share before I go on vacation next week. There are likely several issues, one of which involves camera going to invalid rooms and causing the infamous black frames again. I plan to revisit this in March when I return. In the meantime, please feel free to test and play with this. @lahm86 if you have time and are interested, feel free to make any adjustments to this branch while I'm away :relieved: 

Remaining snagging list:
- [ ] [Inventory item jittering](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2682087037) (present in TR1 develop)
- [ ] [3d pickup animations lateral movement](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2676369456) (present in TR1 develop)
- [x] [Flame blast jittering](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2676378424)
- [x] [Odd camera transition during cutscenes](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2676896570) (present in TR1 develop)
- [x] [Interpolated camera going into walls](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2683460559) (present in TR1 develop)
- [ ] [Cutscene objects with large anim shifts](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2685555738) (present in TR1 develop)
- [x] [Grenades/harpoons not interpolated](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2692184104)

rr- committed:
tr2: add 60 fps

Kappa971 commented:
Small question, why only 60 fps? Same thing for TR1X.

aredfan commented:
Issues (feel free to edit this comment to expand the list):

- [x] Boulder in level 12, room 32. You'll need to activate the boulder and wait for it to stop rolling - 60 fps
- [x] Black screen when starting cutscenes - 30/60 fps
- [x] Play gym, `/tp` to the switch in the maze and press it, the camera shifts a little when it focuses on the door, and shifts again when it returns to Lara - 60 fps

rr- commented:
> Small question, why only 60 fps? Same thing for TR1X.

Because the main loop runs at 30 FPS and is doing both logic and drawing, whereas now we interpolate every other frame. Interpolating the entire rendering would be more complex.

Richard-L commented:
This is so insane... TR2 in 60fps 🤯 

Things I noticed on a quick test play:
- Lara is jittering a little when climbing/descending ladders. Very subtle, mainyl just on the legs.
- Her braid is jittery. Very easy to see as she uses the zip line in the assualt course.
- The underpart of the skidoo is also jittery (see Tibet demo).
- The skidoo motor SFX loops in a different and unexpected way (maybe should this be unrelated?)
- The passport page open animation is not 60fps
- The rotation animation of the sound setting discman mesh looks like it may be 30fps?
- Like in TR1X, the 3d pickup animations lateral movement is 30fps (the rotation is 60fps)
- The Great Wall cutscene doesn't appear to be working. It just zooms out slowly and doesn't show any characters. Unless this is due to me not providing assets, and accidentally messing up the camera?
![20250222_204730_Level_0](https://github.com/user-attachments/assets/378ad80c-d8d6-48f5-898f-884d5f32ee92)

lahm86 commented:
Amazing.

I've attached a patch for hair, shadows, cutscenes and the room portal issue you pointed out on Discord.
[60.diff.txt](https://github.com/user-attachments/files/18925259/60.diff.txt)

The only other things I've spotted that I can't quite work out:
- [x] The boat judders (and Lara when in it), similar to the skidoo
- [x] When Lara hits a springboard or is thrown by the Bartoli boom, she judders. High fall speed issue maybe? Could be the same as #1206.
- [x] I think the dragon and flamethrower flame blasts are jittery (but this could be my eyesight tbh 😄)
- [x] Exploded mines have the same issue as [boulders that have stopped](https://github.com/LostArtefacts/TRX/pull/2544#issuecomment-2676346191)

rr- committed:
interpolation: fix vehicles

rr- committed:
interpolation: fix springboards

lahm86 committed:
items: avoid interpolating specific items

This covers not interpolating items that are inactive, such as stopped
boulders. An item may still be active in the item list such that it
retains collision, but inanimate, sich as boulders or mines, so we need
to check both statuses. The TR1 test here has also been simplified to
match.

TR2 mines have a bad frame in the level data, so we avoid interpolating
these in all cases. The push button also has a single frame where a
mesh is rotated 180deg to simulate off/on, so this can't be
interpolated.

lahm86 committed:
objects/common: move Object_DrawInterpolatedObject to TRX

This moves Object_DrawInterpolatedObject from TR1 to TRX, and updates
TR2's Object_DrawAnimatingItem to use this as the logic is identical.

lahm86 committed:
tr2/inventory_ring/draw: draw interpolated objects

This updates TR2's inventory to use Object_DrawInterpolatedObject so
interpolation is taken into account. Custom behaviour for the stopwatch
retained.

aredfan commented:
- [x] I noticed there's an odd transition for a frame or so when the camera changes position during cutscenes. Here's a few screencaps from cutscene 2 to show you what I mean.

This frame shows very briefly when the camera switches from Lara to Marco.
![1 1](https://github.com/user-attachments/assets/fae98e10-d510-4f31-a901-d13032af9311)
The camera settles here.
![1 2](https://github.com/user-attachments/assets/2ca0d6f3-8bcf-4e6d-a965-4fa959b96be7)
This frame shows very briefly when the camera switches from Marco to the pilot.
![2 1](https://github.com/user-attachments/assets/44accedd-2171-4fb6-8187-99132303a290)
The camera settles here.
![2 2](https://github.com/user-attachments/assets/12ce69ba-78d8-4755-b29d-215bcf27f68d)

lahm86 commented:
@aredfan, @Richard-L Please see #2550 for the latest build (having write issues to this branch). Please continue to track issues/comments here only though so that we just have one reference, the other PR will be closed later.

These problems should be addressed:
- Flickering boulder/exploded mines
- Inventory items, such as passport/CD player

I'll look into the camera problems next.

Richard-L commented:
Looks lovely so far Lahm!

aredfan commented:
> These problems should be addressed:
> 
> * Flickering boulder/exploded mines

@lahm86 I can confirm the boulder issue is fixed, thank you. The mines bug still manifests if you activate the mines first from these triggers:

![Screenshot](https://github.com/user-attachments/assets/0642a6a8-34ce-423c-8627-586255c9f30b)

lahm86 commented:
@aredfan Thank you, good find. It turns out that there is a bad frame in the level data (you can see in WadTool). It never gets processed in 30fps, hence why we're not seeing it here. I've put in a hard-coded check for this until such times as we can fix the bad data.

lahm86 committed:
interpolation: detect large camera shifts

This updates camera interpolation in cases where there are large shifts
of any positional value, such as going to a fixed camera on the other
side of the world. In this case, we just commit the values rather than
interpolating each one, as this can result in the camera having settle
at the destination.

aredfan commented:
> @aredfan Thank you, good find. It turns out that there is a bad frame in the level data (you can see in WadTool). It never gets processed in 30fps, hence why we're not seeing it here. I've put in a hard-coded check for this until such times as we can fix the bad data.

Anytime. I can confirm this issue is now fixed. 👍 

___________________

- [ ] It's possible for objects to jitter when selected in the inventory, it depends on which way they're facing. I think it's an issue in TR1X as well but it's a lot less perceptible there.

**Jitter:**

https://github.com/user-attachments/assets/3f5f2a2b-843b-4647-8ff5-24898a49b40d

**No jitter:**

https://github.com/user-attachments/assets/81322ae5-a55f-422c-85e7-182da0cb015a

lahm86 commented:
I've got a preliminary fix in #2550 for the camera issues, and it'd be good to check where this applies in TR1 as well per #1186. I think it'd be good to check areas such as Temple of the Cat that have lots of panoramic cameras (i.e. gliding big shifts) to ensure there aren't any regressions there.

Edit: noticed a slight issue in my logic, now addressed.

I'll check the inventory issue soon.

aredfan commented:
> I've got a preliminary fix in #2550 for the camera issues, and it'd be good to check where this applies in TR1 as well per #1186. I think it'd be good to check areas such as Temple of the Cat that have lots of panoramic cameras (i.e. gliding big shifts) to ensure there aren't any regressions there.
> 
> Edit: noticed a slight issue in my logic, now addressed.

Well I'm blowed because your solution actually fixed a longstanding issue, in the case of TR1, thank you again. I've checked several of the panoramic cameras in the TRUB Egyptian levels and I don't see any new regressions.

___________________

I found a couple of issues after further testing:

- [x] There's a frame or so of the button disappearing when Lara presses it - 60 fps

![Screenshot 1](https://github.com/user-attachments/assets/410b3b8b-766c-48e5-95f5-aa08ec55b171)

- [x] The gong hammer is dark during the animation where Lara hits the gong - 30/60 fps

Left is the latest dev snapshot, right is the PR build

![Screenshot 2](https://github.com/user-attachments/assets/cb768bfa-06fa-43f0-89c0-dd96b153bb14)

lahm86 committed:
tr2/objects/detonator: fix gong bonger shade

I think prior to this, the shade value happened to be less than zero so
it was having dynamic lighting applied, but the value should be
enforced on creation.

lahm86 commented:
Thanks again, @aredfan. Both the latest issues should now be resolved. The button is interesting, as it has a single frame where the inner mesh is rotated 180deg. So we have to disable interpolation here, because I don't think we can achieve it otherwise (or that it makes sense to). LMK if you think it's behaving OK. It would be good too if you could double check the Venice mines as the logic is bundled.

Re inventory jittering, I'm a bit stuck here. I think I'll add the sticking points to the main PR description above so we have a master snagging list.

lahm86 committed:
interpolation: handle specific effect deltas

This resolves various TR1 and TR2 effects jittering in 60fps, and
ensures grenades and harpoons (from Lara) are interpolated.

lahm86 commented:
Flamethrower/dragon blast should also be resolved now, plus the mutant missiles in TR1 (#1314).

aredfan commented:
@lahm86 the mines, flamethrower, dragon blast, and mutant missiles LGTM.  I think the mutant darts is still 30 fps.

The button issue seems to persist for me on the 0.9.2-66-g711d70b build.

lahm86 commented:
> @lahm86 the mines, flamethrower, dragon blast, and mutant missiles LGTM. I think the mutant darts is still 30 fps.
> 
> The button issue seems to persist for me on the 0.9.2-66-g711d70b build.

Thank you. Button and mutant darts should be fixed now.

I've added another update for the camera too, so LMK if cutscenes are appearing better in both games.

walkawayy commented:
Looks amazing! Only issues I've noticed so far:

1. Like TR1, the roll animation still looks 30 FPS / stuttery.
2. There are some black screen flashes when the camera goes through a wall. Easy to spot if you slide down the opening to Opera House at -2 speed. Idk if this is 60 FPS only bug or not and this may also still have issues in TR1 as well.

![20250225_173714_Opera_House](https://github.com/user-attachments/assets/b035fa19-c876-4fd8-954c-6a20757a5120)

aredfan commented:
> Thank you. Button and mutant darts should be fixed now.

The button issue is fixed, LGTM. 👍 

I'm not sure if the mutant darts are interpolating. I've checked develop as well and the mutant darts look the same there.

__________________

- [ ] I noticed this issue in the first cutscene, the laptop screen does disappear for a frame or so when the screen changes from one texture to another. Pay attention to the gap at the bottom of the laptop's screen. Only 60 fps is affected.

![Screenshot](https://github.com/user-attachments/assets/c2fba06f-9e9c-48cf-b75f-1e4f701b9532)

_________________

- [ ] In the beginning of the 1st cutscene of TR1, the gun lying on the ground visibly shifts when using 60 fps.

lahm86 commented:
Thanks @aredfan. I don't think we can fix the laptop and the gun - it's similar to the push button where the "big shift" happens in the animation, but unlike that, these objects are different in different cutscenes, meaning we can't apply a blanket flag not to interpolate. E.g. the laptop is Bartoli's goon in the Opera cut, the gun is the scion etc. We'd have to introduce the ability to not interpolate objects depending on the level. IMO that's a future issue we could perhaps address.

I'll look at the darts again.

lahm86 commented:
Thanks @walkawayy - the black screen issue is a 60 FPS bug, and it can happen in TR1 as well. In Opera we have this situation. The two outer points are the would-be 30 FPS positions, the middle is the interpolated one. It's in the wall, and so because there's a portal there, the camera room is thrown to the roof, but the position stays below. So the game draws room 42 only for a single frame.

![image](https://github.com/user-attachments/assets/a51f245b-dc92-4d3d-a85d-f2a05235a416)

I think we need to clamp the camera after setting the interpolation results.

lahm86 committed:
interpolation: lower camera delta

This fixes some frames in TR2 cutscenes where the camera perspective
changes, but not quite enough to hit on any coordinate to reach the 512
threshold.

lahm86 commented:
@aredfan hopefully darts are now fixed, along with the TR2 cutscenes from Discord. I also noticed knives and harpoons (from the divers only) were very slightly jittery, so these are fixed too.

aredfan commented:
> @aredfan hopefully darts are now fixed, along with the TR2 cutscenes from Discord. I also noticed knives and harpoons (from the divers only) were very slightly jittery, so these are fixed too.

The mutant darts look much better now, thank you. I could've sworn I was having double vision before the darts got fixed lol.  The knives and harpoons also looks great. 👍 

I can happily confirm that 5 of the odd camera transitions are now fixed.  The last odd camera transition, the # 6 issue from the 2nd cutscene does still persist.  This one is between the first and second time the camera is focused on Bartoli.

aredfan commented:
I can confirm the scene with Bartoli is now fixed 🎉 

I've also re-checked the cutscenes in TR1/2 and all of those LGTM.
The fixed cameras previously brought up in the PR looks good too, in any case, I will keep at it with the sanity check for other parts of gameplay.
Thank you 👍

lahm86 committed:
camera: clamp interpolation result

This performs bounds tests on the interpolated camera result to avoid
instances where black screens can occur.

lahm86 commented:
Pushed a fix for the interpolated camera result sometimes going into walls, so it should fix the Opera House issue @walkawayy. And we're also rebased on latest develop.

walkawayy commented:
It looks really good to me! I didn't notice any issues other than the ones compiled in the PR. Amazing work all!

aredfan commented:
The only remaining issue I could find is the grenades doesn't have interpolation.

Below is a short video, first half is 60 fps and second half is 30 fps, when using negative turbo speeds of -2.

https://github.com/user-attachments/assets/c7c293da-be8b-478b-be52-cb1a4b877280

lahm86 commented:
Thanks @aredfan, should be fixed now. Lara's harpoons I think were suffering from the same, although less clear to see, so resolved too.

aredfan commented:
@lahm86 both grenades and harpoons LGTM. 👍 

I made a video to be sure. Each scenario is 30 fps first then 60 fps second.

https://github.com/user-attachments/assets/c5fb7326-a963-4127-8136-3ee85237b490

rr- reviewed (commented):
Huge thanks to Lahm for driving this home! This LGTM but I can't review PRs opened by myself; IMO we can merge it as-is. WDYT @LostArtefacts/qa @LostArtefacts/dev ?

aredfan reviewed (approved):
> Huge thanks to Lahm for driving this home! This LGTM but I can't review PRs opened by myself; IMO we can merge it as-is. WDYT @LostArtefacts/qa @LostArtefacts/dev ?

I agree, the 60 fps mode for TR2X is fit for purpose, and not to mention several notable bugs afflicting TR1X are now fixed too.

Approval notwithstanding, I only just noticed the TR2X's config tool needs the 30/60 fps text added to the drop-down selection.

rr- committed:
tr2/config-tool: fix missing labels for fps

rr- commented:
@aredfan I think it should be fixed.

aredfan commented:
> @aredfan I think it should be fixed.

Thank you, the config is now showing the fps options as expected. 👍 

I've found another issue where the game reverts to 60 fps despite 30 fps getting selected in the config tool.

**Steps:**
1. launch TR2X and set it to 60 fps
2. close TR2X
3. open the config tool
4. set to 30 fps and save
5. launch TR2X and you'll find it reverts to 60 fps

rr- commented:
> > @aredfan I think it should be fixed.
> 
> Thank you, the config is now showing the fps options as expected. 👍
> 
> I've found another issue where the game reverts to 60 fps despite 30 fps getting selected in the config tool.
> 
> **Steps:**
> 
> 1. launch TR2X and set it to 60 fps
> 2. close TR2X
> 3. open the config tool
> 4. set to 30 fps and save
> 5. launch TR2X and you'll find it reverts to 60 fps

I think that's because we use enums, which internally are stored as strings. I chose enums over normal numeric fields to make it not possible for the user to select "49" for example as the FPS value.

rr- committed:
config/file: allow string integers

The config tool uses enums as the FPS property backing field in order to
limit user choice to either 30 or 60 FPS. The downside is that enum
types get serialized as strings, and TRX wouldn't recognize such values,
expected literal integers. The patch relaxes this requirement and makes
the config parser also accept numeric strings for integer-types.

rr- commented:
@aredfan added a workaround. However, the opposite problem might apply to the config tool now (whereas launching and exiting the game will cause the config tool to revert back to defaults), but I'd leave that alone.

aredfan reviewed (approved):
> @aredfan added a workaround. However, the opposite problem might apply to the config tool now (whereas launching and exiting the game will cause the config tool to revert back to defaults), but I'd leave that alone.

Thank you, LGTM.

The issue with the config tool fps option reverting to default doesn't seem to manifest so that's all good. 👍

rr- committed:
interpolation: fix tr2 demo 3 crashing on exit

The reference in LARA_INFO to the mounted vehicle wasn't cleared when
the demo would finish playing. The interpolation module didn't check if
the Lara object was still valid, and as such, would try to refer to a no
longer existing item which crashed the Windows builds.

rr- committed:
tr1/camera: set the room number for cutscenes

Fixes weird camera angle for cutscene 4. Fix by Lahm.
